{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a65adfeb-03ca-45a0-86be-26029b8fa252",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4db0c87e-31c9-4cc4-b44f-f2c99c3336e9",
   "metadata": {},
   "source": [
    "<h3>Pandas Series</h3>\n",
    "<p>We can build a Pandas Series by using other data structures, such as lists:</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1fd8c6b0-7369-4a17-945a-9f88031219de",
   "metadata": {},
   "outputs": [],
   "source": [
    "value1 = [] #Created an empty list\n",
    "for x in range(5):\n",
    "    value1.append(random.randint(-5,5)) #Appended the random values to the list\n",
    "\n",
    "s1 = pd.Series(value1)\n",
    "s2 = pd.Series(value1, index= [\"a\", \"b\", \"c\", \"d\", \"e\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "478f0959-9675-4fae-b3d4-f18e91c1aebb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "a    0\n",
       "b   -2\n",
       "c    2\n",
       "d    1\n",
       "e    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "241d052b-0d25-4a0b-b1ef-d553272f6ed9",
   "metadata": {},
   "source": [
    "<p>Or Dictionaries</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1bf2ddf4-722c-46cb-9f7e-340fe13a596e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Escherichia coli K-12               4.6\n",
       "Streptomyces coelicolor A3          8.7\n",
       "Bacillus subtilis 168               4.2\n",
       "Pseudomonas aeruginosa PAO1         6.3\n",
       "Mycobacterium tuberculosis H37Rv    4.4\n",
       "dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterialData = {\"Escherichia coli K-12\":4.6,\n",
    "                \"Streptomyces coelicolor A3\":8.7,\n",
    "                \"Bacillus subtilis 168\":4.2,\n",
    "                \"Pseudomonas aeruginosa PAO1\":6.3,\n",
    "                \"Mycobacterium tuberculosis H37Rv\":4.4}\n",
    "\n",
    "bacteriaSeries = pd.Series(bacterialData)\n",
    "bacteriaSeries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cd4c1994-1c2e-4512-b718-9ebc35f25608",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Escherichia coli K-12               False\n",
       "Streptomyces coelicolor A3           True\n",
       "Bacillus subtilis 168               False\n",
       "Pseudomonas aeruginosa PAO1          True\n",
       "Mycobacterium tuberculosis H37Rv    False\n",
       "dtype: bool"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriaSeries >6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8998c4c7-2748-4432-b838-9391fc8d326a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Streptomyces coelicolor A3     8.7\n",
       "Pseudomonas aeruginosa PAO1    6.3\n",
       "dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriaSeries[bacteriaSeries>6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c5d25ce-8944-4e23-b314-b058ab529385",
   "metadata": {},
   "source": [
    "<p>It's easy to add new elements to a series:</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "580c44be-d747-4ddb-a226-c3d7dbe27220",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Escherichia coli K-12               4.6\n",
       "Streptomyces coelicolor A3          8.7\n",
       "Bacillus subtilis 168               4.2\n",
       "Pseudomonas aeruginosa PAO1         6.3\n",
       "Mycobacterium tuberculosis H37Rv    4.4\n",
       "Staphylococcus aureus N315          2.8\n",
       "Helicobacter pylori 26695           1.7\n",
       "dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriaSeries[\"Staphylococcus aureus N315\"] = 2.8\n",
    "bacteriaSeries[\"Helicobacter pylori 26695\"] = 1.7\n",
    "bacteriaSeries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "562e7ad2-5e01-49b9-bdce-bb763117fb0d",
   "metadata": {},
   "source": [
    "<p>Or to modify it:</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "50615492-84a6-4bf3-bab3-0101685056b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Escherichia coli K-12               4.6\n",
       "Streptomyces coelicolor A3          7.0\n",
       "Bacillus subtilis 168               4.2\n",
       "Pseudomonas aeruginosa PAO1         6.3\n",
       "Mycobacterium tuberculosis H37Rv    4.4\n",
       "Staphylococcus aureus N315          2.8\n",
       "Helicobacter pylori 26695           1.7\n",
       "dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriaSeries[\"Streptomyces coelicolor A3\"] = 7\n",
    "bacteriaSeries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d768c089-5d85-4548-81b9-5edb98ae6b65",
   "metadata": {},
   "source": [
    "<p>or:</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "539546fc-0cfb-40e3-ac90-1deb4872068f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Escherichia coli K-12               4.6\n",
       "Streptomyces coelicolor A3          7.0\n",
       "Bacillus subtilis 168               4.2\n",
       "Pseudomonas aeruginosa PAO1         6.3\n",
       "Mycobacterium tuberculosis H37Rv    4.4\n",
       "Staphylococcus aureus N315          3.5\n",
       "Helicobacter pylori 26695           3.5\n",
       "dtype: float64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriaSeries[bacteriaSeries < 3] = 3.5\n",
    "bacteriaSeries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244240e0-5c39-48fd-ba13-09af25a1d087",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fbdc1123-0136-4bc2-b5b2-f9bc363f34e1",
   "metadata": {},
   "source": [
    "<div style=\"padding-top=10px; padding-bottom=10px\">\n",
    "    <h3>DataFrames</h3>\n",
    "    <p><strong>How to create a dataframe:</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d5e91cc2-ec8d-49ee-bdfc-62141084f961",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Genome_size</th>\n",
       "      <th>TranscriptonFactors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces coelicolor A3</td>\n",
       "      <td>8.7</td>\n",
       "      <td>965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Species  Genome_size  TranscriptonFactors\n",
       "0             Escherichia coli K-12          4.6                  314\n",
       "1        Streptomyces coelicolor A3          8.7                  965\n",
       "2             Bacillus subtilis 168          4.2                  158\n",
       "3       Pseudomonas aeruginosa PAO1          6.3                  505\n",
       "4  Mycobacterium tuberculosis H37Rv          4.4                  189"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterialData = {\"Species\": [\"Escherichia coli K-12\",\"Streptomyces coelicolor A3\",\n",
    "                \"Bacillus subtilis 168\",\"Pseudomonas aeruginosa PAO1\",\n",
    "                \"Mycobacterium tuberculosis H37Rv\"],\n",
    "                \"Genome_size\":[4.6,8.7,4.2,6.3,4.4],\n",
    "                \"TranscriptonFactors\":[314,965,158,505,189]}\n",
    "\n",
    "bacterialDataFrame = pd.DataFrame(bacterialData)\n",
    "bacterialDataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d959f908-c2ad-482e-83d2-ef0a6613a80c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5045d05b-1a93-4b0c-ae0b-f67f1b8f6209",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacterium</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Escherichia coli K-12</th>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptomyces coelicolor A3(2)</th>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillus subtilis 168</th>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonas aeruginosa PAO1</th>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium tuberculosis H37Rv</th>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Genome Size (Mbp)  Number of Genes  \\\n",
       "Bacterium                                                              \n",
       "Escherichia coli K-12                           4.6             4400   \n",
       "Streptomyces coelicolor A3(2)                   8.7             7800   \n",
       "Bacillus subtilis 168                           4.2             4100   \n",
       "Pseudomonas aeruginosa PAO1                     6.3             5500   \n",
       "Mycobacterium tuberculosis H37Rv                4.4             4000   \n",
       "\n",
       "                                  Number of Transcription Factors  \\\n",
       "Bacterium                                                           \n",
       "Escherichia coli K-12                                         314   \n",
       "Streptomyces coelicolor A3(2)                                 965   \n",
       "Bacillus subtilis 168                                         158   \n",
       "Pseudomonas aeruginosa PAO1                                   505   \n",
       "Mycobacterium tuberculosis H37Rv                              189   \n",
       "\n",
       "                                  Number of Proteins          Genus  \\\n",
       "Bacterium                                                             \n",
       "Escherichia coli K-12                           4290    Escherichia   \n",
       "Streptomyces coelicolor A3(2)                   7825   Streptomyces   \n",
       "Bacillus subtilis 168                           4103       Bacillus   \n",
       "Pseudomonas aeruginosa PAO1                     5570    Pseudomonas   \n",
       "Mycobacterium tuberculosis H37Rv                4000  Mycobacterium   \n",
       "\n",
       "                                              Family  GC Content (%)  \n",
       "Bacterium                                                             \n",
       "Escherichia coli K-12             Enterobacteriaceae            50.8  \n",
       "Streptomyces coelicolor A3(2)      Streptomycetaceae            72.1  \n",
       "Bacillus subtilis 168                    Bacillaceae            43.5  \n",
       "Pseudomonas aeruginosa PAO1         Pseudomonadaceae            66.6  \n",
       "Mycobacterium tuberculosis H37Rv    Mycobacteriaceae            65.6  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData = pd.read_csv(\"Datasets/bacteria.csv\", index_col=0)\n",
    "fullBacteriaData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6111092c-6fc4-4add-ae39-139e88034eb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "                                       Bacterium  Genome Size (Mbp)  \\\n",
      "0                          Escherichia coli K-12                4.6   \n",
      "1                  Streptomyces coelicolor A3(2)                8.7   \n",
      "2                          Bacillus subtilis 168                4.2   \n",
      "3                    Pseudomonas aeruginosa PAO1                6.3   \n",
      "4               Mycobacterium tuberculosis H37Rv                4.4   \n",
      "..                                           ...                ...   \n",
      "149         Mycobacterium smegmatis str. MC2 155                6.9   \n",
      "150                Mycobacterium bovis AF2122/97                4.3   \n",
      "151  Mycobacterium bovis BCG str. Pasteur 1173P2                4.4   \n",
      "152                  Mycobacterium leprae Br4923                3.3   \n",
      "153           Mycobacterium tuberculosis CDC1551                4.4   \n",
      "\n",
      "     Number of Genes  Number of Transcription Factors  Number of Proteins  \\\n",
      "0               4400                              314                4290   \n",
      "1               7800                              965                7825   \n",
      "2               4100                              158                4103   \n",
      "3               5500                              505                5570   \n",
      "4               4000                              189                4000   \n",
      "..               ...                              ...                 ...   \n",
      "149             7000                              471                7004   \n",
      "150             4000                              257                3997   \n",
      "151             4200                              267                4188   \n",
      "152             1600                               50                1604   \n",
      "153             4000                              257                4004   \n",
      "\n",
      "             Genus              Family  GC Content (%)  \n",
      "0      Escherichia  Enterobacteriaceae            50.8  \n",
      "1     Streptomyces   Streptomycetaceae            72.1  \n",
      "2         Bacillus         Bacillaceae            43.5  \n",
      "3      Pseudomonas    Pseudomonadaceae            66.6  \n",
      "4    Mycobacterium    Mycobacteriaceae            65.6  \n",
      "..             ...                 ...             ...  \n",
      "149  Mycobacterium    Mycobacteriaceae            67.7  \n",
      "150  Mycobacterium    Mycobacteriaceae            65.6  \n",
      "151  Mycobacterium    Mycobacteriaceae            65.5  \n",
      "152  Mycobacterium    Mycobacteriaceae            57.8  \n",
      "153  Mycobacterium    Mycobacteriaceae            65.6  \n",
      "\n",
      "[154 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "fullBacteriaData = pd.read_csv(\"Datasets/bacteria.csv\")\n",
    "print(type(fullBacteriaData))\n",
    "print(fullBacteriaData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "18552abf-171f-4829-9e72-8286a368ef46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>6.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>4.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>3.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Bacterium  Genome Size (Mbp)  \\\n",
       "0                          Escherichia coli K-12                4.6   \n",
       "1                  Streptomyces coelicolor A3(2)                8.7   \n",
       "2                          Bacillus subtilis 168                4.2   \n",
       "3                    Pseudomonas aeruginosa PAO1                6.3   \n",
       "4               Mycobacterium tuberculosis H37Rv                4.4   \n",
       "..                                           ...                ...   \n",
       "149         Mycobacterium smegmatis str. MC2 155                6.9   \n",
       "150                Mycobacterium bovis AF2122/97                4.3   \n",
       "151  Mycobacterium bovis BCG str. Pasteur 1173P2                4.4   \n",
       "152                  Mycobacterium leprae Br4923                3.3   \n",
       "153           Mycobacterium tuberculosis CDC1551                4.4   \n",
       "\n",
       "     Number of Genes  Number of Transcription Factors  Number of Proteins  \\\n",
       "0               4400                              314                4290   \n",
       "1               7800                              965                7825   \n",
       "2               4100                              158                4103   \n",
       "3               5500                              505                5570   \n",
       "4               4000                              189                4000   \n",
       "..               ...                              ...                 ...   \n",
       "149             7000                              471                7004   \n",
       "150             4000                              257                3997   \n",
       "151             4200                              267                4188   \n",
       "152             1600                               50                1604   \n",
       "153             4000                              257                4004   \n",
       "\n",
       "             Genus              Family  GC Content (%)  \n",
       "0      Escherichia  Enterobacteriaceae            50.8  \n",
       "1     Streptomyces   Streptomycetaceae            72.1  \n",
       "2         Bacillus         Bacillaceae            43.5  \n",
       "3      Pseudomonas    Pseudomonadaceae            66.6  \n",
       "4    Mycobacterium    Mycobacteriaceae            65.6  \n",
       "..             ...                 ...             ...  \n",
       "149  Mycobacterium    Mycobacteriaceae            67.7  \n",
       "150  Mycobacterium    Mycobacteriaceae            65.6  \n",
       "151  Mycobacterium    Mycobacteriaceae            65.5  \n",
       "152  Mycobacterium    Mycobacteriaceae            57.8  \n",
       "153  Mycobacterium    Mycobacteriaceae            65.6  \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e74a1421-1d05-42d4-b65d-3a55c7b50c8d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "9aa5a930-bc3e-4d3b-b45a-de0cccbbfd90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>154.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>153.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.504156</td>\n",
       "      <td>3253.896104</td>\n",
       "      <td>204.844156</td>\n",
       "      <td>3257.110390</td>\n",
       "      <td>48.758170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2.092125</td>\n",
       "      <td>1965.543485</td>\n",
       "      <td>156.865086</td>\n",
       "      <td>1964.615083</td>\n",
       "      <td>15.101924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.200000</td>\n",
       "      <td>160.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>161.000000</td>\n",
       "      <td>16.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.825000</td>\n",
       "      <td>1700.000000</td>\n",
       "      <td>91.750000</td>\n",
       "      <td>1713.250000</td>\n",
       "      <td>38.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.350000</td>\n",
       "      <td>3200.000000</td>\n",
       "      <td>187.500000</td>\n",
       "      <td>3205.000000</td>\n",
       "      <td>49.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>4.700000</td>\n",
       "      <td>4400.000000</td>\n",
       "      <td>297.500000</td>\n",
       "      <td>4409.750000</td>\n",
       "      <td>64.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>9.100000</td>\n",
       "      <td>8300.000000</td>\n",
       "      <td>965.000000</td>\n",
       "      <td>8213.000000</td>\n",
       "      <td>72.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Genome Size (Mbp)  Number of Genes  Number of Transcription Factors  \\\n",
       "count         154.000000       154.000000                       154.000000   \n",
       "mean            3.504156      3253.896104                       204.844156   \n",
       "std             2.092125      1965.543485                       156.865086   \n",
       "min             0.200000       160.000000                         1.000000   \n",
       "25%             1.825000      1700.000000                        91.750000   \n",
       "50%             3.350000      3200.000000                       187.500000   \n",
       "75%             4.700000      4400.000000                       297.500000   \n",
       "max             9.100000      8300.000000                       965.000000   \n",
       "\n",
       "       Number of Proteins  GC Content (%)  \n",
       "count          154.000000      153.000000  \n",
       "mean          3257.110390       48.758170  \n",
       "std           1964.615083       15.101924  \n",
       "min            161.000000       16.000000  \n",
       "25%           1713.250000       38.000000  \n",
       "50%           3205.000000       49.500000  \n",
       "75%           4409.750000       64.800000  \n",
       "max           8213.000000       72.100000  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc8038b-f8ca-4e9d-a20d-d9eb906bf176",
   "metadata": {},
   "source": [
    "<h3>Indexing a dataframe</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ab03989b-20a8-4ed7-a700-3dacb248a781",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>6.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>4.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>3.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Bacterium  Genome Size (Mbp)  \\\n",
       "0                          Escherichia coli K-12                4.6   \n",
       "1                  Streptomyces coelicolor A3(2)                8.7   \n",
       "2                          Bacillus subtilis 168                4.2   \n",
       "3                    Pseudomonas aeruginosa PAO1                6.3   \n",
       "4               Mycobacterium tuberculosis H37Rv                4.4   \n",
       "..                                           ...                ...   \n",
       "149         Mycobacterium smegmatis str. MC2 155                6.9   \n",
       "150                Mycobacterium bovis AF2122/97                4.3   \n",
       "151  Mycobacterium bovis BCG str. Pasteur 1173P2                4.4   \n",
       "152                  Mycobacterium leprae Br4923                3.3   \n",
       "153           Mycobacterium tuberculosis CDC1551                4.4   \n",
       "\n",
       "     Number of Genes  Number of Transcription Factors  Number of Proteins  \\\n",
       "0               4400                              314                4290   \n",
       "1               7800                              965                7825   \n",
       "2               4100                              158                4103   \n",
       "3               5500                              505                5570   \n",
       "4               4000                              189                4000   \n",
       "..               ...                              ...                 ...   \n",
       "149             7000                              471                7004   \n",
       "150             4000                              257                3997   \n",
       "151             4200                              267                4188   \n",
       "152             1600                               50                1604   \n",
       "153             4000                              257                4004   \n",
       "\n",
       "             Genus              Family  GC Content (%)  \n",
       "0      Escherichia  Enterobacteriaceae            50.8  \n",
       "1     Streptomyces   Streptomycetaceae            72.1  \n",
       "2         Bacillus         Bacillaceae            43.5  \n",
       "3      Pseudomonas    Pseudomonadaceae            66.6  \n",
       "4    Mycobacterium    Mycobacteriaceae            65.6  \n",
       "..             ...                 ...             ...  \n",
       "149  Mycobacterium    Mycobacteriaceae            67.7  \n",
       "150  Mycobacterium    Mycobacteriaceae            65.6  \n",
       "151  Mycobacterium    Mycobacteriaceae            65.5  \n",
       "152  Mycobacterium    Mycobacteriaceae            57.8  \n",
       "153  Mycobacterium    Mycobacteriaceae            65.6  \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData = pd.read_csv(\"Datasets/bacteria.csv\")\n",
    "fullBacteriaData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "1e8d1522-7ee9-42fe-97a1-91ed3c015380",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome_Size</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Transcription_Factors</th>\n",
       "      <th>Proteins</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Escherichia</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptomyces</th>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillus</th>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonas</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>6.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>4.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>3.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Bacterium  Genome_Size  \\\n",
       "Genus                                                                     \n",
       "Escherichia                          Escherichia coli K-12          4.6   \n",
       "Streptomyces                 Streptomyces coelicolor A3(2)          8.7   \n",
       "Bacillus                             Bacillus subtilis 168          4.2   \n",
       "Pseudomonas                    Pseudomonas aeruginosa PAO1          6.3   \n",
       "Mycobacterium             Mycobacterium tuberculosis H37Rv          4.4   \n",
       "...                                                    ...          ...   \n",
       "Mycobacterium         Mycobacterium smegmatis str. MC2 155          6.9   \n",
       "Mycobacterium                Mycobacterium bovis AF2122/97          4.3   \n",
       "Mycobacterium  Mycobacterium bovis BCG str. Pasteur 1173P2          4.4   \n",
       "Mycobacterium                  Mycobacterium leprae Br4923          3.3   \n",
       "Mycobacterium           Mycobacterium tuberculosis CDC1551          4.4   \n",
       "\n",
       "               Genes  Transcription_Factors  Proteins              Family  \\\n",
       "Genus                                                                       \n",
       "Escherichia     4400                    314      4290  Enterobacteriaceae   \n",
       "Streptomyces    7800                    965      7825   Streptomycetaceae   \n",
       "Bacillus        4100                    158      4103         Bacillaceae   \n",
       "Pseudomonas     5500                    505      5570    Pseudomonadaceae   \n",
       "Mycobacterium   4000                    189      4000    Mycobacteriaceae   \n",
       "...              ...                    ...       ...                 ...   \n",
       "Mycobacterium   7000                    471      7004    Mycobacteriaceae   \n",
       "Mycobacterium   4000                    257      3997    Mycobacteriaceae   \n",
       "Mycobacterium   4200                    267      4188    Mycobacteriaceae   \n",
       "Mycobacterium   1600                     50      1604    Mycobacteriaceae   \n",
       "Mycobacterium   4000                    257      4004    Mycobacteriaceae   \n",
       "\n",
       "                 GC  \n",
       "Genus                \n",
       "Escherichia    50.8  \n",
       "Streptomyces   72.1  \n",
       "Bacillus       43.5  \n",
       "Pseudomonas    66.6  \n",
       "Mycobacterium  65.6  \n",
       "...             ...  \n",
       "Mycobacterium  67.7  \n",
       "Mycobacterium  65.6  \n",
       "Mycobacterium  65.5  \n",
       "Mycobacterium  57.8  \n",
       "Mycobacterium  65.6  \n",
       "\n",
       "[154 rows x 7 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData = pd.read_csv(\"Datasets/bacteria.csv\", index_col=5)\n",
    "fullBacteriaData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "181d5ace-e46f-47a2-9690-3ad971d75176",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>6.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>4.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>3.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Genus                                    Bacterium  \\\n",
       "0      Escherichia                        Escherichia coli K-12   \n",
       "1     Streptomyces                Streptomyces coelicolor A3(2)   \n",
       "2         Bacillus                        Bacillus subtilis 168   \n",
       "3      Pseudomonas                  Pseudomonas aeruginosa PAO1   \n",
       "4    Mycobacterium             Mycobacterium tuberculosis H37Rv   \n",
       "..             ...                                          ...   \n",
       "149  Mycobacterium         Mycobacterium smegmatis str. MC2 155   \n",
       "150  Mycobacterium                Mycobacterium bovis AF2122/97   \n",
       "151  Mycobacterium  Mycobacterium bovis BCG str. Pasteur 1173P2   \n",
       "152  Mycobacterium                  Mycobacterium leprae Br4923   \n",
       "153  Mycobacterium           Mycobacterium tuberculosis CDC1551   \n",
       "\n",
       "     Genome Size (Mbp)  Number of Genes  Number of Transcription Factors  \\\n",
       "0                  4.6             4400                              314   \n",
       "1                  8.7             7800                              965   \n",
       "2                  4.2             4100                              158   \n",
       "3                  6.3             5500                              505   \n",
       "4                  4.4             4000                              189   \n",
       "..                 ...              ...                              ...   \n",
       "149                6.9             7000                              471   \n",
       "150                4.3             4000                              257   \n",
       "151                4.4             4200                              267   \n",
       "152                3.3             1600                               50   \n",
       "153                4.4             4000                              257   \n",
       "\n",
       "     Number of Proteins              Family  GC Content (%)  \n",
       "0                  4290  Enterobacteriaceae            50.8  \n",
       "1                  7825   Streptomycetaceae            72.1  \n",
       "2                  4103         Bacillaceae            43.5  \n",
       "3                  5570    Pseudomonadaceae            66.6  \n",
       "4                  4000    Mycobacteriaceae            65.6  \n",
       "..                  ...                 ...             ...  \n",
       "149                7004    Mycobacteriaceae            67.7  \n",
       "150                3997    Mycobacteriaceae            65.6  \n",
       "151                4188    Mycobacteriaceae            65.5  \n",
       "152                1604    Mycobacteriaceae            57.8  \n",
       "153                4004    Mycobacteriaceae            65.6  \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "1f4303fe-1321-4a53-b8d1-7ecf3d5d80ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Enterobacteriaceae</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptomycetaceae</th>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillaceae</th>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonadaceae</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacteriaceae</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacteriaceae</th>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>6.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacteriaceae</th>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>4.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacteriaceae</th>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacteriaceae</th>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>3.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacteriaceae</th>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Bacterium  \\\n",
       "Family                                                            \n",
       "Enterobacteriaceae                        Escherichia coli K-12   \n",
       "Streptomycetaceae                 Streptomyces coelicolor A3(2)   \n",
       "Bacillaceae                               Bacillus subtilis 168   \n",
       "Pseudomonadaceae                    Pseudomonas aeruginosa PAO1   \n",
       "Mycobacteriaceae               Mycobacterium tuberculosis H37Rv   \n",
       "...                                                         ...   \n",
       "Mycobacteriaceae           Mycobacterium smegmatis str. MC2 155   \n",
       "Mycobacteriaceae                  Mycobacterium bovis AF2122/97   \n",
       "Mycobacteriaceae    Mycobacterium bovis BCG str. Pasteur 1173P2   \n",
       "Mycobacteriaceae                    Mycobacterium leprae Br4923   \n",
       "Mycobacteriaceae             Mycobacterium tuberculosis CDC1551   \n",
       "\n",
       "                    Genome Size (Mbp)  Number of Genes  \\\n",
       "Family                                                   \n",
       "Enterobacteriaceae                4.6             4400   \n",
       "Streptomycetaceae                 8.7             7800   \n",
       "Bacillaceae                       4.2             4100   \n",
       "Pseudomonadaceae                  6.3             5500   \n",
       "Mycobacteriaceae                  4.4             4000   \n",
       "...                               ...              ...   \n",
       "Mycobacteriaceae                  6.9             7000   \n",
       "Mycobacteriaceae                  4.3             4000   \n",
       "Mycobacteriaceae                  4.4             4200   \n",
       "Mycobacteriaceae                  3.3             1600   \n",
       "Mycobacteriaceae                  4.4             4000   \n",
       "\n",
       "                    Number of Transcription Factors  Number of Proteins  \\\n",
       "Family                                                                    \n",
       "Enterobacteriaceae                              314                4290   \n",
       "Streptomycetaceae                               965                7825   \n",
       "Bacillaceae                                     158                4103   \n",
       "Pseudomonadaceae                                505                5570   \n",
       "Mycobacteriaceae                                189                4000   \n",
       "...                                             ...                 ...   \n",
       "Mycobacteriaceae                                471                7004   \n",
       "Mycobacteriaceae                                257                3997   \n",
       "Mycobacteriaceae                                267                4188   \n",
       "Mycobacteriaceae                                 50                1604   \n",
       "Mycobacteriaceae                                257                4004   \n",
       "\n",
       "                    GC Content (%)  \n",
       "Family                              \n",
       "Enterobacteriaceae            50.8  \n",
       "Streptomycetaceae             72.1  \n",
       "Bacillaceae                   43.5  \n",
       "Pseudomonadaceae              66.6  \n",
       "Mycobacteriaceae              65.6  \n",
       "...                            ...  \n",
       "Mycobacteriaceae              67.7  \n",
       "Mycobacteriaceae              65.6  \n",
       "Mycobacteriaceae              65.5  \n",
       "Mycobacteriaceae              57.8  \n",
       "Mycobacteriaceae              65.6  \n",
       "\n",
       "[154 rows x 6 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.set_index(\"Family\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "4f811fa9-3597-4cb8-a3cb-c0d2ad979dd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        Escherichia\n",
       "1       Streptomyces\n",
       "2           Bacillus\n",
       "3        Pseudomonas\n",
       "4      Mycobacterium\n",
       "           ...      \n",
       "149    Mycobacterium\n",
       "150    Mycobacterium\n",
       "151    Mycobacterium\n",
       "152    Mycobacterium\n",
       "153    Mycobacterium\n",
       "Name: Genus, Length: 154, dtype: object"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.reset_index(inplace=True)\n",
    "fullBacteriaData[\"Genus\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "5fd5dea4-afd8-4134-88e6-6d6c7ecee9ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Genus              Family\n",
       "0      Escherichia  Enterobacteriaceae\n",
       "1     Streptomyces   Streptomycetaceae\n",
       "2         Bacillus         Bacillaceae\n",
       "3      Pseudomonas    Pseudomonadaceae\n",
       "4    Mycobacterium    Mycobacteriaceae\n",
       "..             ...                 ...\n",
       "149  Mycobacterium    Mycobacteriaceae\n",
       "150  Mycobacterium    Mycobacteriaceae\n",
       "151  Mycobacterium    Mycobacteriaceae\n",
       "152  Mycobacterium    Mycobacteriaceae\n",
       "153  Mycobacterium    Mycobacteriaceae\n",
       "\n",
       "[154 rows x 2 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData[[\"Genus\", \"Family\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5b9b8066-c93a-40d6-aab3-5e6b15cbf967",
   "metadata": {},
   "outputs": [],
   "source": [
    "fullBacteriaData.reset_index(inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "c04a746f-d65d-47a7-abb6-bce9d2e8452d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Genus                                     Escherichia\n",
       "Genome Size (Mbp)                                 4.6\n",
       "Number of Genes                                  4400\n",
       "Number of Transcription Factors                   314\n",
       "Number of Proteins                               4290\n",
       "Family                             Enterobacteriaceae\n",
       "GC Content (%)                                   50.8\n",
       "Name: Escherichia coli K-12, dtype: object"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.set_index(\"Bacterium\", inplace=True)\n",
    "fullBacteriaData.loc[\"Escherichia coli K-12\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2b99a7e1-cb41-4b67-a976-ac94cf4acd4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Escherichia\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8y/qd2lk9td24q_thh4lq6_6brm0000gr/T/ipykernel_8209/194522097.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  print(fullBacteriaData.loc[\"Escherichia coli K-12\"][0])\n"
     ]
    }
   ],
   "source": [
    "print(fullBacteriaData.loc[\"Escherichia coli K-12\"][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "cb604f29-7e86-49f4-a8f7-e2873948b171",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Genus                                     Escherichia\n",
       "Genome Size (Mbp)                                 4.6\n",
       "Number of Genes                                  4400\n",
       "Number of Transcription Factors                   314\n",
       "Number of Proteins                               4290\n",
       "Family                             Enterobacteriaceae\n",
       "GC Content (%)                                   50.8\n",
       "Name: Escherichia coli K-12, dtype: object"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "a772b0c5-b83b-4c37-b928-382fc809d19e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8y/qd2lk9td24q_thh4lq6_6brm0000gr/T/ipykernel_8209/708634817.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  print(fullBacteriaData.iloc[0][1])\n"
     ]
    }
   ],
   "source": [
    "print(fullBacteriaData.iloc[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "6472fcea-36dc-4d16-8104-fd8104171f3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Genome Size (Mbp)</th>\n",
       "      <th>Number of Genes</th>\n",
       "      <th>Number of Transcription Factors</th>\n",
       "      <th>Number of Proteins</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC Content (%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacterium</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ralstonia solanacearum GMI1000</th>\n",
       "      <td>Ralstonia</td>\n",
       "      <td>5.8</td>\n",
       "      <td>5400</td>\n",
       "      <td>428</td>\n",
       "      <td>5385</td>\n",
       "      <td>Burkholderiaceae</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xanthomonas campestris ATCC 33913</th>\n",
       "      <td>Xanthomonas</td>\n",
       "      <td>5.2</td>\n",
       "      <td>4900</td>\n",
       "      <td>340</td>\n",
       "      <td>4901</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bartonella henselae Houston-1</th>\n",
       "      <td>Bartonella</td>\n",
       "      <td>1.9</td>\n",
       "      <td>1800</td>\n",
       "      <td>48</td>\n",
       "      <td>1754</td>\n",
       "      <td>Bartonellaceae</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Legionella pneumophila Paris</th>\n",
       "      <td>Legionella</td>\n",
       "      <td>3.4</td>\n",
       "      <td>3000</td>\n",
       "      <td>184</td>\n",
       "      <td>2969</td>\n",
       "      <td>Legionellaceae</td>\n",
       "      <td>38.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yersinia pestis CO92</th>\n",
       "      <td>Yersinia</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4200</td>\n",
       "      <td>282</td>\n",
       "      <td>4201</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>47.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Brucella abortus 2308</th>\n",
       "      <td>Brucella</td>\n",
       "      <td>3.3</td>\n",
       "      <td>3200</td>\n",
       "      <td>110</td>\n",
       "      <td>3214</td>\n",
       "      <td>Brucellaceae</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bordetella pertussis Tohama I</th>\n",
       "      <td>Bordetella</td>\n",
       "      <td>4.1</td>\n",
       "      <td>3800</td>\n",
       "      <td>203</td>\n",
       "      <td>3861</td>\n",
       "      <td>Alcaligenaceae</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treponema pallidum Nichols</th>\n",
       "      <td>Treponema</td>\n",
       "      <td>1.1</td>\n",
       "      <td>1000</td>\n",
       "      <td>25</td>\n",
       "      <td>1010</td>\n",
       "      <td>Spirochaetaceae</td>\n",
       "      <td>52.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chlamydia pneumoniae CWL029</th>\n",
       "      <td>Chlamydia</td>\n",
       "      <td>1.2</td>\n",
       "      <td>1100</td>\n",
       "      <td>19</td>\n",
       "      <td>1112</td>\n",
       "      <td>Chlamydiaceae</td>\n",
       "      <td>42.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synechocystis PCC 6803</th>\n",
       "      <td>Synechocystis</td>\n",
       "      <td>3.6</td>\n",
       "      <td>3400</td>\n",
       "      <td>247</td>\n",
       "      <td>3381</td>\n",
       "      <td>Synechococcaceae</td>\n",
       "      <td>47.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Genus  Genome Size (Mbp)  \\\n",
       "Bacterium                                                             \n",
       "Ralstonia solanacearum GMI1000         Ralstonia                5.8   \n",
       "Xanthomonas campestris ATCC 33913    Xanthomonas                5.2   \n",
       "Bartonella henselae Houston-1         Bartonella                1.9   \n",
       "Legionella pneumophila Paris          Legionella                3.4   \n",
       "Yersinia pestis CO92                    Yersinia                4.6   \n",
       "Brucella abortus 2308                   Brucella                3.3   \n",
       "Bordetella pertussis Tohama I         Bordetella                4.1   \n",
       "Treponema pallidum Nichols             Treponema                1.1   \n",
       "Chlamydia pneumoniae CWL029            Chlamydia                1.2   \n",
       "Synechocystis PCC 6803             Synechocystis                3.6   \n",
       "\n",
       "                                   Number of Genes  \\\n",
       "Bacterium                                            \n",
       "Ralstonia solanacearum GMI1000                5400   \n",
       "Xanthomonas campestris ATCC 33913             4900   \n",
       "Bartonella henselae Houston-1                 1800   \n",
       "Legionella pneumophila Paris                  3000   \n",
       "Yersinia pestis CO92                          4200   \n",
       "Brucella abortus 2308                         3200   \n",
       "Bordetella pertussis Tohama I                 3800   \n",
       "Treponema pallidum Nichols                    1000   \n",
       "Chlamydia pneumoniae CWL029                   1100   \n",
       "Synechocystis PCC 6803                        3400   \n",
       "\n",
       "                                   Number of Transcription Factors  \\\n",
       "Bacterium                                                            \n",
       "Ralstonia solanacearum GMI1000                                 428   \n",
       "Xanthomonas campestris ATCC 33913                              340   \n",
       "Bartonella henselae Houston-1                                   48   \n",
       "Legionella pneumophila Paris                                   184   \n",
       "Yersinia pestis CO92                                           282   \n",
       "Brucella abortus 2308                                          110   \n",
       "Bordetella pertussis Tohama I                                  203   \n",
       "Treponema pallidum Nichols                                      25   \n",
       "Chlamydia pneumoniae CWL029                                     19   \n",
       "Synechocystis PCC 6803                                         247   \n",
       "\n",
       "                                   Number of Proteins              Family  \\\n",
       "Bacterium                                                                   \n",
       "Ralstonia solanacearum GMI1000                   5385    Burkholderiaceae   \n",
       "Xanthomonas campestris ATCC 33913                4901    Xanthomonadaceae   \n",
       "Bartonella henselae Houston-1                    1754      Bartonellaceae   \n",
       "Legionella pneumophila Paris                     2969      Legionellaceae   \n",
       "Yersinia pestis CO92                             4201  Enterobacteriaceae   \n",
       "Brucella abortus 2308                            3214        Brucellaceae   \n",
       "Bordetella pertussis Tohama I                    3861      Alcaligenaceae   \n",
       "Treponema pallidum Nichols                       1010     Spirochaetaceae   \n",
       "Chlamydia pneumoniae CWL029                      1112       Chlamydiaceae   \n",
       "Synechocystis PCC 6803                           3381    Synechococcaceae   \n",
       "\n",
       "                                   GC Content (%)  \n",
       "Bacterium                                          \n",
       "Ralstonia solanacearum GMI1000               67.0  \n",
       "Xanthomonas campestris ATCC 33913            65.0  \n",
       "Bartonella henselae Houston-1                39.0  \n",
       "Legionella pneumophila Paris                 38.4  \n",
       "Yersinia pestis CO92                         47.6  \n",
       "Brucella abortus 2308                        57.0  \n",
       "Bordetella pertussis Tohama I                67.0  \n",
       "Treponema pallidum Nichols                   52.8  \n",
       "Chlamydia pneumoniae CWL029                  42.1  \n",
       "Synechocystis PCC 6803                       47.7  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index_list = list(range(20,30))\n",
    "\n",
    "fullBacteriaData.iloc[index_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "dd460fd1-9491-478f-9892-e9401a8f018d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[20, 21, 22, 23, 24, 25, 26, 27, 28, 29]\n"
     ]
    }
   ],
   "source": [
    "print(index_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a5e54d1d-bc3c-4f78-97fb-1b9d08d895eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacterium\n",
      "Escherichia coli K-12                 Escherichia\n",
      "Streptomyces coelicolor A3(2)        Streptomyces\n",
      "Bacillus subtilis 168                    Bacillus\n",
      "Pseudomonas aeruginosa PAO1           Pseudomonas\n",
      "Mycobacterium tuberculosis H37Rv    Mycobacterium\n",
      "Name: Genus, dtype: object\n",
      "Bacterium\n",
      "Mycobacterium smegmatis str. MC2 155           Mycobacterium\n",
      "Mycobacterium bovis AF2122/97                  Mycobacterium\n",
      "Mycobacterium bovis BCG str. Pasteur 1173P2    Mycobacterium\n",
      "Mycobacterium leprae Br4923                    Mycobacterium\n",
      "Mycobacterium tuberculosis CDC1551             Mycobacterium\n",
      "Name: Genus, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(fullBacteriaData[\"Genus\"].head())\n",
    "print(fullBacteriaData[\"Genus\"].tail())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "bde16b91-46fc-4ead-850a-0c4f0aba43dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacterium\n",
      "Mycobacterium vanbaalenii PYR-1                Mycobacterium\n",
      "Mycobacterium gilvum PYR-GCK                   Mycobacterium\n",
      "Mycobacterium smegmatis str. MC2 155           Mycobacterium\n",
      "Mycobacterium bovis AF2122/97                  Mycobacterium\n",
      "Mycobacterium bovis BCG str. Pasteur 1173P2    Mycobacterium\n",
      "Mycobacterium leprae Br4923                    Mycobacterium\n",
      "Mycobacterium tuberculosis CDC1551             Mycobacterium\n",
      "Name: Genus, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(fullBacteriaData[\"Genus\"].tail(7))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "c5e665b6-c79e-44fb-a78a-5aadb0c35349",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Genus\n",
       "Mycobacterium             26\n",
       "Helicobacter               4\n",
       "Candidatus Blochmannia     3\n",
       "Leptospira                 3\n",
       "Burkholderia               3\n",
       "Mycoplasma                 3\n",
       "Bacteroides                3\n",
       "Streptococcus              3\n",
       "Nitrospira                 3\n",
       "Escherichia                2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData[\"Genus\"].value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "4a578cbc-5ba2-494f-ae43-c93a815374c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome_Size</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Transcription_Factors</th>\n",
       "      <th>Proteins</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Escherichia</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>4.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptomyces</th>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>8.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillus</th>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>4.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonas</th>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>6.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>4.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>3.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium</th>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>4.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Bacterium  Genome_Size  \\\n",
       "Genus                                                                     \n",
       "Escherichia                          Escherichia coli K-12          4.6   \n",
       "Streptomyces                 Streptomyces coelicolor A3(2)          8.7   \n",
       "Bacillus                             Bacillus subtilis 168          4.2   \n",
       "Pseudomonas                    Pseudomonas aeruginosa PAO1          6.3   \n",
       "Mycobacterium             Mycobacterium tuberculosis H37Rv          4.4   \n",
       "...                                                    ...          ...   \n",
       "Mycobacterium         Mycobacterium smegmatis str. MC2 155          6.9   \n",
       "Mycobacterium                Mycobacterium bovis AF2122/97          4.3   \n",
       "Mycobacterium  Mycobacterium bovis BCG str. Pasteur 1173P2          4.4   \n",
       "Mycobacterium                  Mycobacterium leprae Br4923          3.3   \n",
       "Mycobacterium           Mycobacterium tuberculosis CDC1551          4.4   \n",
       "\n",
       "               Genes  Transcription_Factors  Proteins              Family  \\\n",
       "Genus                                                                       \n",
       "Escherichia     4400                    314      4290  Enterobacteriaceae   \n",
       "Streptomyces    7800                    965      7825   Streptomycetaceae   \n",
       "Bacillus        4100                    158      4103         Bacillaceae   \n",
       "Pseudomonas     5500                    505      5570    Pseudomonadaceae   \n",
       "Mycobacterium   4000                    189      4000    Mycobacteriaceae   \n",
       "...              ...                    ...       ...                 ...   \n",
       "Mycobacterium   7000                    471      7004    Mycobacteriaceae   \n",
       "Mycobacterium   4000                    257      3997    Mycobacteriaceae   \n",
       "Mycobacterium   4200                    267      4188    Mycobacteriaceae   \n",
       "Mycobacterium   1600                     50      1604    Mycobacteriaceae   \n",
       "Mycobacterium   4000                    257      4004    Mycobacteriaceae   \n",
       "\n",
       "                 GC  \n",
       "Genus                \n",
       "Escherichia    50.8  \n",
       "Streptomyces   72.1  \n",
       "Bacillus       43.5  \n",
       "Pseudomonas    66.6  \n",
       "Mycobacterium  65.6  \n",
       "...             ...  \n",
       "Mycobacterium  67.7  \n",
       "Mycobacterium  65.6  \n",
       "Mycobacterium  65.5  \n",
       "Mycobacterium  57.8  \n",
       "Mycobacterium  65.6  \n",
       "\n",
       "[154 rows x 7 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3109e553-c44b-43d6-920a-3e292f076085",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "24bc3767-d80c-4f99-b746-e6943ebd0799",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The average GC content of the dataset is 48.76%\n",
      "Genus\n",
      "Escherichia      4.6\n",
      "Streptomyces     8.7\n",
      "Bacillus         4.2\n",
      "Pseudomonas      6.3\n",
      "Mycobacterium    4.4\n",
      "Name: Genome_Size, dtype: float64\n",
      "Genus\n",
      "Escherichia       9.6\n",
      "Streptomyces     13.7\n",
      "Bacillus          9.2\n",
      "Pseudomonas      11.3\n",
      "Mycobacterium     9.4\n",
      "Name: Genome_Size, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "averageGC = fullBacteriaData[\"GC\"].mean()\n",
    "print(f\"The average GC content of the dataset is {round(averageGC,2)}%\")\n",
    "\n",
    "GenomeSize = fullBacteriaData[\"Genome_Size\"]\n",
    "print(GenomeSize.head(5))\n",
    "GenomeSize += 5\n",
    "print(GenomeSize.head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1676b467-e210-46a9-be0d-0615f2012802",
   "metadata": {},
   "source": [
    "<h3>GroupBy:</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "17bbabe1-0fbd-4dc6-ac35-c6337a9007ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome_Size</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Transcription_Factors</th>\n",
       "      <th>Proteins</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>9.6</td>\n",
       "      <td>4400</td>\n",
       "      <td>314</td>\n",
       "      <td>4290</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>50.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>13.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus</td>\n",
       "      <td>Bacillus subtilis 168</td>\n",
       "      <td>9.2</td>\n",
       "      <td>4100</td>\n",
       "      <td>158</td>\n",
       "      <td>4103</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>11.3</td>\n",
       "      <td>5500</td>\n",
       "      <td>505</td>\n",
       "      <td>5570</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>66.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>9.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>189</td>\n",
       "      <td>4000</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium smegmatis str. MC2 155</td>\n",
       "      <td>11.9</td>\n",
       "      <td>7000</td>\n",
       "      <td>471</td>\n",
       "      <td>7004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium bovis AF2122/97</td>\n",
       "      <td>9.3</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>3997</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium bovis BCG str. Pasteur 1173P2</td>\n",
       "      <td>9.4</td>\n",
       "      <td>4200</td>\n",
       "      <td>267</td>\n",
       "      <td>4188</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium leprae Br4923</td>\n",
       "      <td>8.3</td>\n",
       "      <td>1600</td>\n",
       "      <td>50</td>\n",
       "      <td>1604</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>57.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium tuberculosis CDC1551</td>\n",
       "      <td>9.4</td>\n",
       "      <td>4000</td>\n",
       "      <td>257</td>\n",
       "      <td>4004</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>65.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Genus                                    Bacterium  Genome_Size  \\\n",
       "0      Escherichia                        Escherichia coli K-12          9.6   \n",
       "1     Streptomyces                Streptomyces coelicolor A3(2)         13.7   \n",
       "2         Bacillus                        Bacillus subtilis 168          9.2   \n",
       "3      Pseudomonas                  Pseudomonas aeruginosa PAO1         11.3   \n",
       "4    Mycobacterium             Mycobacterium tuberculosis H37Rv          9.4   \n",
       "..             ...                                          ...          ...   \n",
       "149  Mycobacterium         Mycobacterium smegmatis str. MC2 155         11.9   \n",
       "150  Mycobacterium                Mycobacterium bovis AF2122/97          9.3   \n",
       "151  Mycobacterium  Mycobacterium bovis BCG str. Pasteur 1173P2          9.4   \n",
       "152  Mycobacterium                  Mycobacterium leprae Br4923          8.3   \n",
       "153  Mycobacterium           Mycobacterium tuberculosis CDC1551          9.4   \n",
       "\n",
       "     Genes  Transcription_Factors  Proteins              Family    GC  \n",
       "0     4400                    314      4290  Enterobacteriaceae  50.8  \n",
       "1     7800                    965      7825   Streptomycetaceae  72.1  \n",
       "2     4100                    158      4103         Bacillaceae  43.5  \n",
       "3     5500                    505      5570    Pseudomonadaceae  66.6  \n",
       "4     4000                    189      4000    Mycobacteriaceae  65.6  \n",
       "..     ...                    ...       ...                 ...   ...  \n",
       "149   7000                    471      7004    Mycobacteriaceae  67.7  \n",
       "150   4000                    257      3997    Mycobacteriaceae  65.6  \n",
       "151   4200                    267      4188    Mycobacteriaceae  65.5  \n",
       "152   1600                     50      1604    Mycobacteriaceae  57.8  \n",
       "153   4000                    257      4004    Mycobacteriaceae  65.6  \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData#.groupby(\"Genus\").groups.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "245cd92b-cdd6-4e52-bd1e-6e0391096d56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Family\n",
       "Actinomycetaceae        66.0\n",
       "Aeromonadaceae          61.0\n",
       "Alcaligenaceae          67.0\n",
       "Anaplasmataceae         34.0\n",
       "Aquificaceae            43.0\n",
       "                        ... \n",
       "Thermodesulfobiaceae    61.0\n",
       "Thermoplasmataceae      54.0\n",
       "Thermotogaceae          46.2\n",
       "Vibrionaceae            47.7\n",
       "Xanthomonadaceae        65.0\n",
       "Name: GC, Length: 73, dtype: float64"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.groupby(\"Family\")[\"GC\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "e711de56-6f69-4e41-a8b9-ea4024d77a50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Family                Bacterium                                           \n",
       "Actinomycetaceae      Actinomyces naeslundii MG1                              67.0\n",
       "                      Actinomyces odontolyticus ATCC 17982                    65.0\n",
       "Aeromonadaceae        Aeromonas hydrophila subsp. hydrophila ATCC 7966        61.0\n",
       "Alcaligenaceae        Bordetella pertussis Tohama I                           67.0\n",
       "Anaplasmataceae       Wolbachia endosymbiont of Culex quinquefasciatus Pel    34.0\n",
       "                                                                              ... \n",
       "Thermodesulfobiaceae  Thermodesulfobium narugense DSM 14796                   61.0\n",
       "Thermoplasmataceae    Thermoplasma acidophilum DSM 1728                       54.0\n",
       "Thermotogaceae        Thermotoga maritima MSB8                                46.2\n",
       "Vibrionaceae          Vibrio cholerae N16961                                  47.7\n",
       "Xanthomonadaceae      Xanthomonas campestris ATCC 33913                       65.0\n",
       "Name: GC, Length: 143, dtype: float64"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullBacteriaData.groupby([\"Family\", \"Bacterium\"])[\"GC\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "111fc1cc-7250-4db0-b24d-53f5ffe2ba54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0      False\n",
      "1       True\n",
      "2      False\n",
      "3       True\n",
      "4       True\n",
      "       ...  \n",
      "149     True\n",
      "150     True\n",
      "151     True\n",
      "152    False\n",
      "153     True\n",
      "Name: GC, Length: 154, dtype: bool\n",
      "             Genus                                    Bacterium  Genome_Size  \\\n",
      "1     Streptomyces                Streptomyces coelicolor A3(2)         13.7   \n",
      "3      Pseudomonas                  Pseudomonas aeruginosa PAO1         11.3   \n",
      "4    Mycobacterium             Mycobacterium tuberculosis H37Rv          9.4   \n",
      "16     Rhodobacter                Rhodobacter sphaeroides 2.4.1          9.6   \n",
      "20       Ralstonia               Ralstonia solanacearum GMI1000         10.8   \n",
      "..             ...                                          ...          ...   \n",
      "148  Mycobacterium                 Mycobacterium gilvum PYR-GCK         11.1   \n",
      "149  Mycobacterium         Mycobacterium smegmatis str. MC2 155         11.9   \n",
      "150  Mycobacterium                Mycobacterium bovis AF2122/97          9.3   \n",
      "151  Mycobacterium  Mycobacterium bovis BCG str. Pasteur 1173P2          9.4   \n",
      "153  Mycobacterium           Mycobacterium tuberculosis CDC1551          9.4   \n",
      "\n",
      "     Genes  Transcription_Factors  Proteins             Family    GC  \n",
      "1     7800                    965      7825  Streptomycetaceae  72.1  \n",
      "3     5500                    505      5570   Pseudomonadaceae  66.6  \n",
      "4     4000                    189      4000   Mycobacteriaceae  65.6  \n",
      "16    4200                    175      4193   Rhodobacteraceae  68.8  \n",
      "20    5400                    428      5385   Burkholderiaceae  67.0  \n",
      "..     ...                    ...       ...                ...   ...  \n",
      "148   5700                    375      5701   Mycobacteriaceae  67.0  \n",
      "149   7000                    471      7004   Mycobacteriaceae  67.7  \n",
      "150   4000                    257      3997   Mycobacteriaceae  65.6  \n",
      "151   4200                    267      4188   Mycobacteriaceae  65.5  \n",
      "153   4000                    257      4004   Mycobacteriaceae  65.6  \n",
      "\n",
      "[52 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "print(fullBacteriaData[\"GC\"]>60)\n",
    "\n",
    "highGC = fullBacteriaData[fullBacteriaData[\"GC\"]>60]\n",
    "print(highGC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "964e7301-1fe0-410d-ae68-dd1a43e0f89e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Bacterium</th>\n",
       "      <th>Genome_Size</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Transcription_Factors</th>\n",
       "      <th>Proteins</th>\n",
       "      <th>Family</th>\n",
       "      <th>GC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Bradyrhizobium</td>\n",
       "      <td>Bradyrhizobium japonicum USDA 110</td>\n",
       "      <td>14.1</td>\n",
       "      <td>8300</td>\n",
       "      <td>621</td>\n",
       "      <td>8213</td>\n",
       "      <td>Bradyrhizobiaceae</td>\n",
       "      <td>64.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Nostoc</td>\n",
       "      <td>Nostoc punctiforme PCC 73102</td>\n",
       "      <td>14.0</td>\n",
       "      <td>8100</td>\n",
       "      <td>590</td>\n",
       "      <td>8023</td>\n",
       "      <td>Nostocaceae</td>\n",
       "      <td>41.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>Streptomyces coelicolor A3(2)</td>\n",
       "      <td>13.7</td>\n",
       "      <td>7800</td>\n",
       "      <td>965</td>\n",
       "      <td>7825</td>\n",
       "      <td>Streptomycetaceae</td>\n",
       "      <td>72.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>Burkholderia cenocepacia AU 1054</td>\n",
       "      <td>13.1</td>\n",
       "      <td>7500</td>\n",
       "      <td>524</td>\n",
       "      <td>7473</td>\n",
       "      <td>Burkholderiaceae</td>\n",
       "      <td>66.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Rhizobium</td>\n",
       "      <td>Rhizobium leguminosarum bv. viciae 3841</td>\n",
       "      <td>12.4</td>\n",
       "      <td>7000</td>\n",
       "      <td>420</td>\n",
       "      <td>7055</td>\n",
       "      <td>Rhizobiaceae</td>\n",
       "      <td>60.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Genus                                Bacterium  Genome_Size  \\\n",
       "46  Bradyrhizobium        Bradyrhizobium japonicum USDA 110         14.1   \n",
       "30          Nostoc             Nostoc punctiforme PCC 73102         14.0   \n",
       "1     Streptomyces            Streptomyces coelicolor A3(2)         13.7   \n",
       "95    Burkholderia         Burkholderia cenocepacia AU 1054         13.1   \n",
       "31       Rhizobium  Rhizobium leguminosarum bv. viciae 3841         12.4   \n",
       "\n",
       "    Genes  Transcription_Factors  Proteins             Family    GC  \n",
       "46   8300                    621      8213  Bradyrhizobiaceae  64.1  \n",
       "30   8100                    590      8023        Nostocaceae  41.5  \n",
       "1    7800                    965      7825  Streptomycetaceae  72.1  \n",
       "95   7500                    524      7473   Burkholderiaceae  66.3  \n",
       "31   7000                    420      7055       Rhizobiaceae  60.5  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorting = fullBacteriaData.sort_values([\"Genome_Size\"], ascending = False)\n",
    "\n",
    "sorting.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe520edb-17a6-492f-989d-7017e9870a5f",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h3>Saving the data into a file</h3>\n",
    "    <p>To save the information of only three columns (Genus, Genome size and GC content) to a text file, we can use the to_csv() method as follows:</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "581d65d9-4526-4912-ab4b-54415acebcc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "shortDataframe = fullBacteriaData[[\"Genus\", \"Genome_Size\", \"GC\"]]\n",
    "\n",
    "shortDataframe.to_csv(\"Datasets/bacteria_short.csv\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7e9c54a7-6ce7-43a2-89d8-290819a14aea",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h3>Concatenate, merge and join</h3>\n",
    "    <p>In any real world data science situation you’ll have to merge or concatenate Dataframes together to manipulate, filter or generate your dataset.</p>\n",
    "    <p>Merging, joining and concatenating are core processes that any aspiring computational biologist will need to master.</p>\n",
    "    <p>Pandas provides methods to combine Series or DataFrame with weird stuff, ranging from Boolean logic to relational algebra to perform join and merge operations</p>\n",
    "    <h3>Concatenating:</h3>\n",
    "    <p>The concat() function will concatenate stuff along an axis while performing optional set logic (union or intersection) of the indexes on the other axes. \n",
    "</p>\n",
    "    <p><center><img src=\"Figures/concat.png\"></center></p>\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3f3e3505-22b0-4de1-a451-4cf0e6dbbb74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    A   B   C   D\n",
       "0  A0  B0  C0  D0\n",
       "1  A1  B1  C1  D1\n",
       "2  A2  B2  C2  D2\n",
       "3  A3  B3  C3  D3"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#First, let's create a mock dataset:\n",
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "        \"A\": [\"A0\", \"A1\", \"A2\", \"A3\"],\n",
    "        \"B\": [\"B0\", \"B1\", \"B2\", \"B3\"],\n",
    "        \"C\": [\"C0\", \"C1\", \"C2\", \"C3\"],\n",
    "        \"D\": [\"D0\", \"D1\", \"D2\", \"D3\"],\n",
    "    },\n",
    "    index = [0,1,2,3],\n",
    ")\n",
    "\n",
    "df2 = pd.DataFrame(\n",
    "    {\n",
    "        \"A\": [\"A4\", \"A5\", \"A6\", \"A7\"],\n",
    "        \"B\": [\"B4\", \"B5\", \"B6\", \"B7\"],\n",
    "        \"C\": [\"C4\", \"C5\", \"C6\", \"C7\"],\n",
    "        \"D\": [\"D4\", \"D5\", \"D6\", \"D7\"],\n",
    "    },\n",
    "    index = [4,5,6,7],\n",
    ")\n",
    "df1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cc4bc07e-0d04-4dde-8203-df5f5cbaaa23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4</td>\n",
       "      <td>B4</td>\n",
       "      <td>C4</td>\n",
       "      <td>D4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A5</td>\n",
       "      <td>B5</td>\n",
       "      <td>C5</td>\n",
       "      <td>D5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A6</td>\n",
       "      <td>B6</td>\n",
       "      <td>C6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A7</td>\n",
       "      <td>B7</td>\n",
       "      <td>C7</td>\n",
       "      <td>D7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    A   B   C   D\n",
       "4  A4  B4  C4  D4\n",
       "5  A5  B5  C5  D5\n",
       "6  A6  B6  C6  D6\n",
       "7  A7  B7  C7  D7"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "722b66e2-5287-4967-93af-c7b5dcc1dae4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4</td>\n",
       "      <td>B4</td>\n",
       "      <td>C4</td>\n",
       "      <td>D4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A5</td>\n",
       "      <td>B5</td>\n",
       "      <td>C5</td>\n",
       "      <td>D5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A6</td>\n",
       "      <td>B6</td>\n",
       "      <td>C6</td>\n",
       "      <td>D6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A7</td>\n",
       "      <td>B7</td>\n",
       "      <td>C7</td>\n",
       "      <td>D7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    A   B   C   D\n",
       "0  A0  B0  C0  D0\n",
       "1  A1  B1  C1  D1\n",
       "2  A2  B2  C2  D2\n",
       "3  A3  B3  C3  D3\n",
       "4  A4  B4  C4  D4\n",
       "5  A5  B5  C5  D5\n",
       "6  A6  B6  C6  D6\n",
       "7  A7  B7  C7  D7"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now let's concatenate\n",
    "\n",
    "pd.concat([df1, df2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566ccb1e-a9f4-40b6-8201-61ae3e2ade6d",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33e68f47-aea2-4161-98f5-5f0174c341c5",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0a19958-b9ef-479b-8cc3-d537b2f029e1",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73d4bc1a-ce01-464f-b6f3-5c0faf810063",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h3>Merging:</h3>\n",
    "    <p>Merging is a convenient method for combining the columns of two Dataframes.</p>\n",
    "    <p>In Pandas, there are separate “merge” and “join” functions, both of which do similar things.</p>\n",
    "    <p>“Merging” two datasets is the process of bringing two datasets together into one, and aligning the rows from each based on common attributes or columns.</p>\n",
    "    <p>“Joining” combines two dataframes on the basis of their indexes.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac18d70-d815-46e3-94d7-3733efdf0fc2",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;margin:auto;\">\n",
    "    <h4>There are four different types of merges available in Pandas:</h4>\n",
    "    <div style=\"display:flex;flex-direction:row; padding-left:10%;\">\n",
    "    <ul style=\"width:20%;\">\n",
    "        <li>Inner merge</li>\n",
    "        <li>Outer merge</li>\n",
    "        <li>Right merge</li>\n",
    "        <li>Left merge</li>\n",
    "    </ul>\n",
    "        <p><img src=\"Figures/merging.png\" style=\"width:70%;\"></p>\n",
    "    </div>\n",
    "    <p>The merge type to use is specified using the “how” parameter in the merge command, taking values “left”, “right”, “inner” (default), or “outer”.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "29e1ef14-7b3a-4675-9b87-92a114d6d42e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K0</td>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K1</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K2</td>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K3</td>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K4</td>\n",
       "      <td>A4</td>\n",
       "      <td>B4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K5</td>\n",
       "      <td>A5</td>\n",
       "      <td>B5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  key   A   B\n",
       "0  K0  A0  B0\n",
       "1  K1  A1  B1\n",
       "2  K2  A2  B2\n",
       "3  K3  A3  B3\n",
       "4  K4  A4  B4\n",
       "5  K5  A5  B5"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame(\n",
    "    {\n",
    "        \"key\": [\"K0\", \"K1\", \"K2\", \"K3\", \"K4\", \"K5\"],\n",
    "        \"A\": [\"A0\", \"A1\", \"A2\", \"A3\", \"A4\", \"A5\"],\n",
    "        \"B\": [\"B0\", \"B1\", \"B2\", \"B3\", \"B4\", \"B5\"],\n",
    "    },\n",
    "    index = [0,1,2,3,4,5],\n",
    ")\n",
    "\n",
    "df2 = pd.DataFrame(\n",
    "    {\n",
    "        \"key\": [\"K0\", \"K1\", \"K2\", \"K3\"],\n",
    "        \"C\": [\"C0\", \"C1\", \"C2\", \"C3\"],\n",
    "        \"D\": [\"D0\", \"D1\", \"D2\", \"D3\"],\n",
    "    },\n",
    "    index = [6,7,8,9],\n",
    ")\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94d8e134-9490-41d4-924c-232ba30b9992",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>K2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>K3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  key   C   D\n",
       "4  K0  C0  D0\n",
       "5  K1  C1  D1\n",
       "6  K2  C2  D2\n",
       "7  K3  C3  D3"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "61519cf0-46d8-46c6-b5a4-7d6e7e517950",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K0</td>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K1</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K2</td>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K3</td>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  key   A   B   C   D\n",
       "0  K0  A0  B0  C0  D0\n",
       "1  K1  A1  B1  C1  D1\n",
       "2  K2  A2  B2  C2  D2\n",
       "3  K3  A3  B3  C3  D3"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Inner merge\n",
    "pd.merge(df1, df2, how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f57a1236-48f3-4e11-a66b-c5ed25988edc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K0</td>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K1</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K2</td>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K3</td>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K4</td>\n",
       "      <td>A4</td>\n",
       "      <td>B4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K5</td>\n",
       "      <td>A5</td>\n",
       "      <td>B5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  key   A   B    C    D\n",
       "0  K0  A0  B0   C0   D0\n",
       "1  K1  A1  B1   C1   D1\n",
       "2  K2  A2  B2   C2   D2\n",
       "3  K3  A3  B3   C3   D3\n",
       "4  K4  A4  B4  NaN  NaN\n",
       "5  K5  A5  B5  NaN  NaN"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#outer merge\n",
    "pd.merge(df1, df2, how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "519b0434-9d36-44a9-a902-3414d078ab5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K0</td>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K1</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K2</td>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K3</td>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K4</td>\n",
       "      <td>A4</td>\n",
       "      <td>B4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K5</td>\n",
       "      <td>A5</td>\n",
       "      <td>B5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  key   A   B    C    D\n",
       "0  K0  A0  B0   C0   D0\n",
       "1  K1  A1  B1   C1   D1\n",
       "2  K2  A2  B2   C2   D2\n",
       "3  K3  A3  B3   C3   D3\n",
       "4  K4  A4  B4  NaN  NaN\n",
       "5  K5  A5  B5  NaN  NaN"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#left merge\n",
    "pd.merge(df1, df2, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1ace89a0-6b56-45c4-81e9-b03a3e846049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K0</td>\n",
       "      <td>A0</td>\n",
       "      <td>B0</td>\n",
       "      <td>C0</td>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K1</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C1</td>\n",
       "      <td>D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K2</td>\n",
       "      <td>A2</td>\n",
       "      <td>B2</td>\n",
       "      <td>C2</td>\n",
       "      <td>D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K3</td>\n",
       "      <td>A3</td>\n",
       "      <td>B3</td>\n",
       "      <td>C3</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  key   A   B   C   D\n",
       "0  K0  A0  B0  C0  D0\n",
       "1  K1  A1  B1  C1  D1\n",
       "2  K2  A2  B2  C2  D2\n",
       "3  K3  A3  B3  C3  D3"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#right merge\n",
    "pd.merge(df1, df2, how='right')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28f25a3b-762e-415d-bad3-6f396acb2db1",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h3>In Pandas, missing Data is represented by two values:</h3>\n",
    "    <p><strong>None</strong>:a Python singleton often used for missing data in Python code.</p>\n",
    "    <p><strong>NaN (Not a number)</strong>:A floating point value recognized by many systems.</p>\n",
    "    <p>They are virtually the same, they mean that there is a value is missing.</p>\n",
    "    <h4>Using isna()/isnull():</h4>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61276139-158e-4684-b2e9-3300cc5dcd35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>First Score</th>\n",
       "      <th>Second Score</th>\n",
       "      <th>Third Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   First Score  Second Score  Third Score\n",
       "0        False         False         True\n",
       "1        False         False        False\n",
       "2         True         False        False\n",
       "3        False          True        False"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "dict = {'First Score':[100,90,np.nan,95],\n",
    "        'Second Score':[30,45,56,np.nan],\n",
    "        'Third Score':[np.nan,40,80,98]}\n",
    "\n",
    "df = pd.DataFrame(dict)\n",
    "df.isna()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0171cc6-e653-40b0-b40a-f02545012ee4",
   "metadata": {},
   "source": [
    "<h4>Using notna()/notnull()</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7bad3d5e-abcf-4d1d-a09e-4c8618aa56fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>First Score</th>\n",
       "      <th>Second Score</th>\n",
       "      <th>Third Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   First Score  Second Score  Third Score\n",
       "0         True          True        False\n",
       "1         True          True         True\n",
       "2        False          True         True\n",
       "3         True         False         True"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.notna()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbef4e5c-98de-4ce8-9f28-ff647c96dbe5",
   "metadata": {},
   "source": [
    "<h4>Using dropna():</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2c746b01-5825-4a47-a22c-321ff1fa0e08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   First Score  Second Score  Third Score\n",
      "0        100.0          30.0          NaN\n",
      "1         90.0          45.0         40.0\n",
      "2          NaN          56.0         80.0\n",
      "3         95.0           NaN         98.0\n",
      "\n",
      "   First Score  Second Score  Third Score\n",
      "1         90.0          45.0         40.0\n",
      "\n",
      "   First Score  Second Score  Third Score\n",
      "0        100.0          30.0          NaN\n",
      "1         90.0          45.0         40.0\n",
      "2          NaN          56.0         80.0\n",
      "3         95.0           NaN         98.0\n"
     ]
    }
   ],
   "source": [
    "print(df)\n",
    "print(\"\")\n",
    "print(df.dropna())\n",
    "print(\"\")\n",
    "print(df.dropna(how=\"all\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b09d41e-800d-4818-a7a0-33afabc25465",
   "metadata": {},
   "source": [
    "<h4>Using fillna():</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "23ffe07e-fe4b-46ae-b37b-29daea3b8397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>First Score</th>\n",
       "      <th>Second Score</th>\n",
       "      <th>Third Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>90.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.5</td>\n",
       "      <td>56.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>95.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>98.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   First Score  Second Score  Third Score\n",
       "0        100.0          30.0          0.5\n",
       "1         90.0          45.0         40.0\n",
       "2          0.5          56.0         80.0\n",
       "3         95.0           0.5         98.0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.fillna(0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f980c96-4a71-4390-bfae-6f47cb699871",
   "metadata": {},
   "source": [
    "<h4>Using replace():</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bbbd22f4-696b-46b1-9e02-eac9e822b96d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>First Score</th>\n",
       "      <th>Second Score</th>\n",
       "      <th>Third Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>90.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.5</td>\n",
       "      <td>56.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>95.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>98.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   First Score  Second Score  Third Score\n",
       "0        100.0          30.0          0.5\n",
       "1         90.0          45.0         40.0\n",
       "2          0.5          56.0         80.0\n",
       "3         95.0           0.5         98.0"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.replace(to_replace = np.nan, value = 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf492cdc-3b17-4a2e-bf80-63ada986bff9",
   "metadata": {},
   "source": [
    "<p><center><img src=\"Figures/dropna.png\"></center></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8aeff068-dc5a-416a-b7ea-9aef25d5c5ee",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h2>Data Standardization:</h2>\n",
    "    <p>Data standardization is the data quality process of transforming data to fit a predefined and constrained set of values.</p>\n",
    "    <p>Basically, we change the scale to compare datapoints from different samples.</p>\n",
    "    <p>It is crucial for quantitative analysis for methodologies such as Transcriptomics, Metabolomics and Metatranscriptomics, as well as Genomics and Metagenomics.</p>\n",
    "    <p>Examples in Biology are:</p>\n",
    "    <ul>\n",
    "        <li><strong>RPKM</strong></li>\n",
    "        <li><strong>FPKM</strong></li>\n",
    "        <li><strong>TPM</strong></li>\n",
    "        <li><strong>RPKG</strong></li>\n",
    "        <li><strong>CODA</strong></li>\n",
    "    </ul>\n",
    "    <p>Since our time is short, we will focus on the CODA part!</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e1b4c73-8739-4a9f-81cc-96167cf7d4c8",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h2>Data Standardization:</h2>\n",
    "    <p>Data standardization is the data quality process of transforming data to fit a predefined and constrained set of values.</p>\n",
    "    <p>Basically, we change the scale to compare datapoints from different samples.</p>\n",
    "    <p>It is crucial for quantitative analysis for methodologies such as Transcriptomics, Metabolomics and Metatranscriptomics, as well as Genomics and Metagenomics.</p>\n",
    "    <p>Examples in Biology are:</p>\n",
    "    <ul>\n",
    "        <li><strong>RPKM</strong></li>\n",
    "        <li><strong>FPKM</strong></li>\n",
    "        <li><strong>TPM</strong></li>\n",
    "        <li><strong>RPKG</strong></li>\n",
    "        <li><strong>CODA</strong></li>\n",
    "    </ul>\n",
    "    <p>Since our time is short, we will focus on the CODA part!</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb00eaf5-235e-43f2-90b1-a52eac61a85d",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h3>CODA:</h3>\n",
    "    <p>16s is limited when compared to Deep sequencing samples, but it has its own value.</p>\n",
    "    <p>However, 16s counts doesn’t make any sense if they do not consider all counts.</p>\n",
    "    <p>That’s why we normally transform 16s data into fractions or percentages.</p>\n",
    "    <p><center><img src = \"Figures/barplot.png\" style=\"width:30%;\"></center></p>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a1984fb-3a0d-47fb-a7f3-4030accf824d",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;margin:auto;\">\n",
    "    <h4>Linear transformations:</h4>\n",
    "    <div style=\"display:flex;flex-direction:row;\">\n",
    "    <div>\n",
    "        <h5>Three log ratio transformations available:</h5>\n",
    "    <ul>\n",
    "        <li>Additive log-ratio (ALR)</li>\n",
    "        <li>Centred log-ration (CLR)</li>\n",
    "        <li>Isometric log-ration (ILR)</li>\n",
    "    </ul>\n",
    "        <h5>All three satisfies linearity, but they are not the same thing.</h5>\n",
    "    <ul>\n",
    "        <li>Sub compositional coherence</li>\n",
    "        <li>Scale invariance</li>\n",
    "    </ul>\n",
    "    </div>\n",
    "        <div style=\"padding-left:10%;\">\n",
    "        <p><img src=\"Figures/linear_transformation.png\" style=\"width:70%;\"></p>\n",
    "    </div>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec59cd11-ab65-4c45-bc8d-bb0ca3de2fcb",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <h3>ALR:</h3>\n",
    "    <p>ALR is one of the simplest transformations.</p>\n",
    "    <p>You will arbitrarily choose a composition common to all samples (maybe the one with less variation) as reference (ref).</p>\n",
    "    <p>You will divide all the other compositions by this particular composition and calculate the log.</p>\n",
    "    <p><center><img src = \"Figures/alr.png\" style=\"width:30%;\"></center></p>\n",
    "    \n",
    "</div>\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a7b0aa86-412b-42cd-b960-c9f9f4db7e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "import skbio\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "microbiomeDF = pd.read_table(\"Datasets/Bacteria_counts.tsv\", index_col=0)\n",
    "taxonomy = pd.read_table(\"Datasets/Bacteria_taxonomy.tsv\", index_col=0)\n",
    "metadata = pd.read_table(\"Datasets/Metadata.tsv\", index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7bdc1859-9e17-48d6-96fc-d8d263799675",
   "metadata": {},
   "outputs": [],
   "source": [
    "def composition(CountsDataFrame):\n",
    "    \"\"\"\n",
    "    Transform the counts table into compositions:\n",
    "\n",
    "    Args:\n",
    "    _____\n",
    "    DataFrame (int): Dataframe containing count values\n",
    "    options:\n",
    "    freq - Transform the table into frequencies\n",
    "    perc - Transform the table into percentages\n",
    "\n",
    "    Return:\n",
    "    _______\n",
    "        Dataframe containing compositional data\n",
    "    \"\"\"\n",
    "\n",
    "    CountsDataFrame = CountsDataFrame.replace(0,1)\n",
    "    if validate_counts(CountsDataFrame) ==False:\n",
    "        raise DataFrameValidationError\n",
    "    else:\n",
    "        return(CountsDataFrame.div(CountsDataFrame.sum(axis=0))*100)\n",
    "\n",
    "def validate_counts(DataFrame):\n",
    "    \"\"\"\n",
    "    Validate the input to check if the sum of components is\n",
    "    \n",
    "    different from a pre-defined upper-limit (1 or 100).\n",
    "    Args:\n",
    "    _____\n",
    "    DataFrame (int): Dataframe containing count values\n",
    "\n",
    "    Return:\n",
    "    _______\n",
    "        Boolean variable\n",
    "    \"\"\"\n",
    "    return(all(round(DataFrame.sum(),1)!=1.0) and all(DataFrame.sum()!=100.0))\n",
    "\n",
    "def validate_composition(DataFrame):\n",
    "    \"\"\"\n",
    "    Validate the input to check if the sum of components is 1 or 100.\n",
    "    \n",
    "    Args:\n",
    "    _____\n",
    "    DataFrame (float): Dataframe containing compositional data\n",
    "\n",
    "    Return:\n",
    "    _______\n",
    "        Boolean variable\n",
    "    \"\"\"\n",
    "    return(all(round(DataFrame.sum(),1)==1.0) or all(round(DataFrame.sum(),1)==100.0))\n",
    "\n",
    "\n",
    "def clr(DataFrame):\n",
    "    \"\"\"\n",
    "    Transform the compositional table into Centered log-ratios\n",
    "    CLR = log(Composition/geometric mean)\n",
    "\n",
    "    Args:\n",
    "    _____\n",
    "    DataFrame (float): Dataframe containing compositional data.\n",
    "\n",
    "    Return:\n",
    "    _______\n",
    "        CLR transformed dataframe\n",
    "    \"\"\"\n",
    "    if validate_composition(DataFrame) ==False:\n",
    "        raise DataFrameValidationError\n",
    "    else:\n",
    "        gmeans = np.log(DataFrame).mean()\n",
    "        clr = np.log(DataFrame)-gmeans\n",
    "        return(clr)\n",
    "\n",
    "def alr(DataFrame, ref):\n",
    "    \"\"\"\n",
    "    Transform the compositional table into Aitchison log-ratios    \n",
    "    ALR = log(Composition/reference)\n",
    "\n",
    "    Args:\n",
    "    _____\n",
    "    DataFrame (float): Dataframe containing compositional data.\n",
    "    ref (str): Composition being used as reference for the log-ratio.\n",
    "    Return:\n",
    "    _______\n",
    "        ALR transformed dataframe\n",
    "    \"\"\"\n",
    "    if validate_composition(DataFrame) ==False:\n",
    "        raise DataFrameValidationError\n",
    "    else:\n",
    "        reference = DataFrame.T[ref]\n",
    "        alr = np.log(DataFrame.T.div(reference, axis=0))\n",
    "        return(alr.drop(ref,axis=1).T)\n",
    "\n",
    "\n",
    "def Aitchison_Distance(Dataframe):\n",
    "    \"\"\"\n",
    "    Calculate the euclidian distance between samples based on CLR transformed data.\n",
    "\n",
    "    Args:\n",
    "    _____\n",
    "    DataFrame (float): Dataframe containing clr-transformed data.\n",
    "    \n",
    "    Return:\n",
    "    _______\n",
    "    Dataframe containing the euclidian distance between samples.\n",
    "    \"\"\"\n",
    "    aitchison_distance = squareform(pdist(Dataframe))\n",
    "    aitchison_distance = pd.DataFrame(aitchison_distance, index=clr_asv.index, columns=clr_asv.index)\n",
    "    return aitchison_distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "fa6d2a68-4dd9-4352-aac2-3f32f536ab43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1462</th>\n",
       "      <th>1464</th>\n",
       "      <th>1479</th>\n",
       "      <th>1485</th>\n",
       "      <th>1494</th>\n",
       "      <th>1495</th>\n",
       "      <th>1496</th>\n",
       "      <th>1498</th>\n",
       "      <th>1501</th>\n",
       "      <th>1503</th>\n",
       "      <th>...</th>\n",
       "      <th>500</th>\n",
       "      <th>502</th>\n",
       "      <th>505</th>\n",
       "      <th>506</th>\n",
       "      <th>510</th>\n",
       "      <th>511</th>\n",
       "      <th>514</th>\n",
       "      <th>515</th>\n",
       "      <th>517</th>\n",
       "      <th>523</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASV_990</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-2.443521</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.305325</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-3.255832</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-3.391988</td>\n",
       "      <td>-2.942488</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-2.728786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_1676</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-4.211916</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-5.471430</td>\n",
       "      <td>-1.643205</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-5.422009</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_1533</th>\n",
       "      <td>-3.237370</td>\n",
       "      <td>-5.059214</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-4.116479</td>\n",
       "      <td>-2.763800</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-5.482951</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-3.915351</td>\n",
       "      <td>-3.486355</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-4.446565</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-5.134327</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_3142</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-1.706963</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-3.497769</td>\n",
       "      <td>-4.429493</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-3.247047</td>\n",
       "      <td>-4.735737</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-6.238325</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_3117</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-6.238325</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_2087</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-3.392579</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-6.238325</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_2574</th>\n",
       "      <td>-5.252273</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-6.061790</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-6.387721</td>\n",
       "      <td>-3.530274</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-2.759421</td>\n",
       "      <td>-5.590987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_4458</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-3.405111</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_3992</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.195697</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-6.238325</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_2614</th>\n",
       "      <td>-5.945421</td>\n",
       "      <td>-7.767264</td>\n",
       "      <td>-5.420535</td>\n",
       "      <td>-2.564949</td>\n",
       "      <td>-7.671827</td>\n",
       "      <td>-6.675823</td>\n",
       "      <td>-7.725330</td>\n",
       "      <td>-6.054439</td>\n",
       "      <td>-6.291569</td>\n",
       "      <td>-7.680176</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.498282</td>\n",
       "      <td>-8.007700</td>\n",
       "      <td>-6.400257</td>\n",
       "      <td>-6.782192</td>\n",
       "      <td>-6.642487</td>\n",
       "      <td>-7.080868</td>\n",
       "      <td>-6.238325</td>\n",
       "      <td>-4.672829</td>\n",
       "      <td>-6.520621</td>\n",
       "      <td>-6.284134</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2782 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              1462      1464      1479      1485      1494      1495  \\\n",
       "ASV_990  -5.945421 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "ASV_1676 -5.945421 -4.211916 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "ASV_1533 -3.237370 -5.059214 -5.420535 -2.564949 -4.116479 -2.763800   \n",
       "ASV_3142 -5.945421 -7.767264 -1.706963 -2.564949 -7.671827 -3.497769   \n",
       "ASV_3117 -5.945421 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "ASV_2087 -5.945421 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "ASV_2574 -5.252273 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "ASV_4458 -5.945421 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "ASV_3992 -5.945421 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "ASV_2614 -5.945421 -7.767264 -5.420535 -2.564949 -7.671827 -6.675823   \n",
       "\n",
       "              1496      1498      1501      1503  ...       500       502  \\\n",
       "ASV_990  -7.725330 -2.443521 -6.291569 -7.680176  ... -1.305325 -8.007700   \n",
       "ASV_1676 -7.725330 -6.054439 -6.291569 -7.680176  ... -6.498282 -8.007700   \n",
       "ASV_1533 -7.725330 -6.054439 -6.291569 -5.482951  ... -6.498282 -8.007700   \n",
       "ASV_3142 -4.429493 -6.054439 -3.247047 -4.735737  ... -6.498282 -8.007700   \n",
       "ASV_3117 -7.725330 -6.054439 -6.291569 -7.680176  ... -6.498282 -8.007700   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "ASV_2087 -7.725330 -6.054439 -6.291569 -7.680176  ... -6.498282 -3.392579   \n",
       "ASV_2574 -7.725330 -6.054439 -6.291569 -7.680176  ... -6.498282 -6.061790   \n",
       "ASV_4458 -7.725330 -6.054439 -6.291569 -7.680176  ... -6.498282 -8.007700   \n",
       "ASV_3992 -7.725330 -6.054439 -6.291569 -7.680176  ... -4.195697 -8.007700   \n",
       "ASV_2614 -7.725330 -6.054439 -6.291569 -7.680176  ... -6.498282 -8.007700   \n",
       "\n",
       "               505       506       510       511       514       515  \\\n",
       "ASV_990  -6.400257 -3.255832 -6.642487 -3.391988 -2.942488 -4.672829   \n",
       "ASV_1676 -6.400257 -6.782192 -6.642487 -5.471430 -1.643205 -4.672829   \n",
       "ASV_1533 -3.915351 -3.486355 -6.642487 -7.080868 -4.446565 -4.672829   \n",
       "ASV_3142 -6.400257 -6.782192 -6.642487 -7.080868 -6.238325 -4.672829   \n",
       "ASV_3117 -6.400257 -6.782192 -6.642487 -7.080868 -6.238325 -4.672829   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "ASV_2087 -6.400257 -6.782192 -6.642487 -7.080868 -6.238325 -4.672829   \n",
       "ASV_2574 -6.400257 -6.782192 -6.642487 -6.387721 -3.530274 -4.672829   \n",
       "ASV_4458 -6.400257 -6.782192 -6.642487 -7.080868 -3.405111 -4.672829   \n",
       "ASV_3992 -6.400257 -6.782192 -6.642487 -7.080868 -6.238325 -4.672829   \n",
       "ASV_2614 -6.400257 -6.782192 -6.642487 -7.080868 -6.238325 -4.672829   \n",
       "\n",
       "               517       523  \n",
       "ASV_990  -6.520621 -2.728786  \n",
       "ASV_1676 -5.422009 -6.284134  \n",
       "ASV_1533 -5.134327 -6.284134  \n",
       "ASV_3142 -6.520621 -6.284134  \n",
       "ASV_3117 -6.520621 -6.284134  \n",
       "...            ...       ...  \n",
       "ASV_2087 -6.520621 -6.284134  \n",
       "ASV_2574 -2.759421 -5.590987  \n",
       "ASV_4458 -6.520621 -6.284134  \n",
       "ASV_3992 -6.520621 -6.284134  \n",
       "ASV_2614 -6.520621 -6.284134  \n",
       "\n",
       "[2782 rows x 52 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_composition = composition(microbiomeDF)\n",
    "alr_transformed = alr(df_composition, \"ASV_3\")\n",
    "alr_transformed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19db440a-47b8-44d1-a70b-1f7f3277b8c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b082c27-437a-4dba-b1e9-e2e090484dc5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ed8d159a-6476-48f1-b250-80bb7c004116",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <p><center><img src = \"Figures/ALR_crow.png\" style=\"width:50%;\"></center></p>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b12779ee-58e9-4169-9d8d-6396e345ab50",
   "metadata": {},
   "source": [
    "<h3>CLR</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "cd8e19a6-493b-4605-aad9-18b8eadf698c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1462</th>\n",
       "      <th>1464</th>\n",
       "      <th>1479</th>\n",
       "      <th>1485</th>\n",
       "      <th>1494</th>\n",
       "      <th>1495</th>\n",
       "      <th>1496</th>\n",
       "      <th>1498</th>\n",
       "      <th>1501</th>\n",
       "      <th>1503</th>\n",
       "      <th>...</th>\n",
       "      <th>500</th>\n",
       "      <th>502</th>\n",
       "      <th>505</th>\n",
       "      <th>506</th>\n",
       "      <th>510</th>\n",
       "      <th>511</th>\n",
       "      <th>514</th>\n",
       "      <th>515</th>\n",
       "      <th>517</th>\n",
       "      <th>523</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ASV_990</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>3.052450</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>4.279574</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>2.900090</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>2.762573</td>\n",
       "      <td>2.281283</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>2.751314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_1676</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>2.888153</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>0.683131</td>\n",
       "      <td>3.580566</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>0.030706</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_1533</th>\n",
       "      <td>2.055518</td>\n",
       "      <td>2.040856</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>2.864250</td>\n",
       "      <td>2.970029</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>1.582088</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>1.337435</td>\n",
       "      <td>2.669566</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>0.777206</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>0.318388</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_3142</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>2.952226</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>2.236059</td>\n",
       "      <td>2.273572</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>2.008879</td>\n",
       "      <td>2.329302</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>-1.014554</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_3117</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>-1.014554</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_2087</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>3.534074</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>-1.014554</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_2574</th>\n",
       "      <td>0.040615</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>0.864863</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.233160</td>\n",
       "      <td>1.693496</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>2.693293</td>\n",
       "      <td>-0.110887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_4458</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>1.818660</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_3992</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>1.389202</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>-1.014554</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASV_2614</th>\n",
       "      <td>-0.652532</td>\n",
       "      <td>-0.667195</td>\n",
       "      <td>-0.761346</td>\n",
       "      <td>-0.679041</td>\n",
       "      <td>-0.691098</td>\n",
       "      <td>-0.941994</td>\n",
       "      <td>-1.022265</td>\n",
       "      <td>-0.558468</td>\n",
       "      <td>-1.035643</td>\n",
       "      <td>-0.615137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.913383</td>\n",
       "      <td>-1.081047</td>\n",
       "      <td>-1.147471</td>\n",
       "      <td>-0.626271</td>\n",
       "      <td>-0.743095</td>\n",
       "      <td>-0.926307</td>\n",
       "      <td>-1.014554</td>\n",
       "      <td>-0.824084</td>\n",
       "      <td>-1.067907</td>\n",
       "      <td>-0.804034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2783 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              1462      1464      1479      1485      1494      1495  \\\n",
       "ASV_990  -0.652532 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "ASV_1676 -0.652532  2.888153 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "ASV_1533  2.055518  2.040856 -0.761346 -0.679041  2.864250  2.970029   \n",
       "ASV_3142 -0.652532 -0.667195  2.952226 -0.679041 -0.691098  2.236059   \n",
       "ASV_3117 -0.652532 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "ASV_2087 -0.652532 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "ASV_2574  0.040615 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "ASV_4458 -0.652532 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "ASV_3992 -0.652532 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "ASV_2614 -0.652532 -0.667195 -0.761346 -0.679041 -0.691098 -0.941994   \n",
       "\n",
       "              1496      1498      1501      1503  ...       500       502  \\\n",
       "ASV_990  -1.022265  3.052450 -1.035643 -0.615137  ...  4.279574 -1.081047   \n",
       "ASV_1676 -1.022265 -0.558468 -1.035643 -0.615137  ... -0.913383 -1.081047   \n",
       "ASV_1533 -1.022265 -0.558468 -1.035643  1.582088  ... -0.913383 -1.081047   \n",
       "ASV_3142  2.273572 -0.558468  2.008879  2.329302  ... -0.913383 -1.081047   \n",
       "ASV_3117 -1.022265 -0.558468 -1.035643 -0.615137  ... -0.913383 -1.081047   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "ASV_2087 -1.022265 -0.558468 -1.035643 -0.615137  ... -0.913383  3.534074   \n",
       "ASV_2574 -1.022265 -0.558468 -1.035643 -0.615137  ... -0.913383  0.864863   \n",
       "ASV_4458 -1.022265 -0.558468 -1.035643 -0.615137  ... -0.913383 -1.081047   \n",
       "ASV_3992 -1.022265 -0.558468 -1.035643 -0.615137  ...  1.389202 -1.081047   \n",
       "ASV_2614 -1.022265 -0.558468 -1.035643 -0.615137  ... -0.913383 -1.081047   \n",
       "\n",
       "               505       506       510       511       514       515  \\\n",
       "ASV_990  -1.147471  2.900090 -0.743095  2.762573  2.281283 -0.824084   \n",
       "ASV_1676 -1.147471 -0.626271 -0.743095  0.683131  3.580566 -0.824084   \n",
       "ASV_1533  1.337435  2.669566 -0.743095 -0.926307  0.777206 -0.824084   \n",
       "ASV_3142 -1.147471 -0.626271 -0.743095 -0.926307 -1.014554 -0.824084   \n",
       "ASV_3117 -1.147471 -0.626271 -0.743095 -0.926307 -1.014554 -0.824084   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "ASV_2087 -1.147471 -0.626271 -0.743095 -0.926307 -1.014554 -0.824084   \n",
       "ASV_2574 -1.147471 -0.626271 -0.743095 -0.233160  1.693496 -0.824084   \n",
       "ASV_4458 -1.147471 -0.626271 -0.743095 -0.926307  1.818660 -0.824084   \n",
       "ASV_3992 -1.147471 -0.626271 -0.743095 -0.926307 -1.014554 -0.824084   \n",
       "ASV_2614 -1.147471 -0.626271 -0.743095 -0.926307 -1.014554 -0.824084   \n",
       "\n",
       "               517       523  \n",
       "ASV_990  -1.067907  2.751314  \n",
       "ASV_1676  0.030706 -0.804034  \n",
       "ASV_1533  0.318388 -0.804034  \n",
       "ASV_3142 -1.067907 -0.804034  \n",
       "ASV_3117 -1.067907 -0.804034  \n",
       "...            ...       ...  \n",
       "ASV_2087 -1.067907 -0.804034  \n",
       "ASV_2574  2.693293 -0.110887  \n",
       "ASV_4458 -1.067907 -0.804034  \n",
       "ASV_3992 -1.067907 -0.804034  \n",
       "ASV_2614 -1.067907 -0.804034  \n",
       "\n",
       "[2783 rows x 52 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clr_transformed = clr(df_composition)\n",
    "clr_transformed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be604919-dbf2-45af-9ffc-154737c66924",
   "metadata": {},
   "source": [
    "<div style=\"padding-top:10px; padding-bottom:10px;\">\n",
    "    <p><center><img src = \"Figures/CLR_CROW.png\" style=\"width:50%;\"></center></p>\n",
    "    \n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
